From fa831bc8c67bddd7537facde6893ac91e0430675 Mon Sep 17 00:00:00 2001
From: kelson8 <kelson@kelsoncraft.net>
Date: Sat, 5 Sep 2026 01:42:34 -0400
Subject: [PATCH 082/161] feat: Make create_player work in free roam script

* Make project use freeroam-game.lua for init in new freeroam mode.

* Add SetupFreeroamScript function to load the freeroam script on init.

* Setup basic player spawning in freeroam-game.lua.

* Fix createPlayerLua function in lua_player.cpp.

* Make lua init work again if DISABLE_GAME_SCRIPTS isn't enabled.

* Add freeroam script to defines.cpp
* Update changelog.md, Bump version to 1.2.6a.
---
 Changelog.md                                  | 17 ++++
 .../lua_scripts/freeroam-game.lua             | 88 +++++++++++++++++++
 src/control/Script.cpp                        | 10 ++-
 src/core/Game.cpp                             |  3 +-
 src/extras/lua/lua_player.cpp                 | 10 ++-
 src/extras/lua/lua_test.cpp                   | 29 +++++-
 src/extras/lua/lua_test.h                     |  1 +
 src/extras/utils/defines.cpp                  |  5 +-
 src/extras/utils/defines.h                    |  1 +
 src/skel/glfw/glfw.cpp                        | 11 +++
 10 files changed, 165 insertions(+), 10 deletions(-)

diff --git a/Changelog.md b/Changelog.md
index 3dff7a67..ed600ebf 100644
--- a/Changelog.md
+++ b/Changelog.md
@@ -58,3 +58,20 @@ feat: Add test for spawning player in lua scripts
 
 * Add CreatePlayer function into MiscFunctions.
 * Update changelog.md, bump version to 1.2.5a.
+
+### 1.2.6a
+
+feat: Make create_player work in free roam script
+
+* Make project use freeroam-game.lua for init in new freeroam mode.
+
+* Add SetupFreeroamScript function to load the freeroam script on init.
+
+* Setup basic player spawning in freeroam-game.lua.
+
+* Fix createPlayerLua function in lua_player.cpp.
+
+* Make lua init work again if DISABLE_GAME_SCRIPTS isn't enabled.
+
+* Add freeroam script to defines.cpp
+* Update changelog.md, Bump version to 1.2.6a.
diff --git a/gamefiles/ViceExtended/lua_scripts/freeroam-game.lua b/gamefiles/ViceExtended/lua_scripts/freeroam-game.lua
index 455c7004..9ad64114 100644
--- a/gamefiles/ViceExtended/lua_scripts/freeroam-game.lua
+++ b/gamefiles/ViceExtended/lua_scripts/freeroam-game.lua
@@ -17,10 +17,34 @@
 -- REQUIRED
 -- If you remove these functions out of the freeroam script you will most likely crash!
 -- They are needed for the game to know what to do without the '.scm' scripts. 
+-- This is pretty much a basic replacement for the game scripts.
 --------
 
 -- Init player, spawn them in.
 
+-----------
+-- Locations
+-----------
+-- TODO Figure out how to move these into another file
+
+local spawnX = -258.1
+local spawnY = -453.1
+local spawnZ = 13.5
+
+local airportX = -1515.1
+local airportY = -1182.1
+local airportZ = 16.5
+local airportHeading = 0.0
+
+-- Spawn in the player, mostly for spawning without the .scm scripts
+-- This works in here now!
+
+-------
+-- REQUIRED if DISABLE_GAME_SCRIPTS is toggled on in the game code, otherwise it won't spawn the player.
+-------
+-- create_player(0, spawnX, spawnY, spawnZ)
+create_player(0, airportX, airportY, airportZ)
+
 --------
 -- End REQUIRED
 --------
@@ -29,6 +53,70 @@
 -- Custom functions go below
 --------
 
+local enableNeverWanted = false
+local enableInfiniteHealth = false
+
+-- Set ped density to 0.0
+local disablePeds = false
+-- Set vehicle density to 0.0
+local disableVehicles = false
+
+local blowUpCarsToggle = false
+
+local clearAreaToggle = false
+
+-- If this is disabled, you won't lose weapons when busted or wasted.
+-- local loseWeapons = false
+local loseWeapons = true
+
+-----------
+-- Toggles
+-----------
+
+-- Toggle losing weapons on death.
+if not loseWeapons then
+	disable_lose_weapons_on_death()
+else
+	enable_lose_weapons_on_death()
+end
+
+
+-- Blow up all cars cheat
+if blowUpCarsToggle then
+	blow_up_all_vehicles()
+end
+
+-- Clear the area of any peds and vehicles.
+if clearAreaToggle then
+	clear_area(25)
+end
+
+-- Turn the ped density multiplier to 0.0
+if disablePeds then
+	disablePeds()
+end
+
+-- Turn the vehicle density multiplier to 0.0
+if disableVehicles then
+	disable_vehicles()
+end
+
+if enableNeverWanted then
+	enable_never_wanted()
+end
+
+if enableInfiniteHealth then
+	enable_infinite_health()
+end
+
+-----------
+-- End Toggles
+-----------
+
 -- Create some vehicles
+-- TODO Fix this in here, this doesn't work just yet.
+-- create_vehicle(145, true, true)
+
 
 -- Spawn some peds
+-- TODO Implement this.
diff --git a/src/control/Script.cpp b/src/control/Script.cpp
index 1243a67b..ba616149 100644
--- a/src/control/Script.cpp
+++ b/src/control/Script.cpp
@@ -840,12 +840,14 @@ void CTheScripts::Init()
 #ifdef DISABLE_GAME_SCRIPTS
 	// Run the lua tick system.
 	// TODO Move this elsewhere, for testing.
-	// LuaTest::Getnstance().OnTick();
+	// LuaTest::GetInstance().OnTick();
 
 	// This works!!!!
-	// I just have to make it work in lua, this is fully bypassing the game scripts but I don't know about it loading my init script.
-	// If I disable this, it will segfault.
-	MiscFunctions::GetInstance().CreatePlayer(0, -258.1f, -453.1f, 13.5f);
+
+	// TODO Make this get an OnInit function in lua, which will create the player and set their position.
+	
+	// I got this moved into the InitLua now which runs under Game.cpp.
+	// MiscFunctions::GetInstance().CreatePlayer(0, -258.1f, -453.1f, 13.5f);
 
 #endif // DISABLE_GAME_SCRIPTS - Switch scm langauge to my lua script testing.
 
diff --git a/src/core/Game.cpp b/src/core/Game.cpp
index b0804162..ea023776 100644
--- a/src/core/Game.cpp
+++ b/src/core/Game.cpp
@@ -628,7 +628,8 @@ bool CGame::Initialise(const char* datFile)
 		CTheScripts::Process();
 #else
 	// Init lua scripts, moved out of glfw.cpp.
-	LuaTest::GetInstance().InitLua();
+	// LuaTest::GetInstance().InitLua();
+	LuaTest::GetInstance().SetupFreeroamScript();
 #endif // !DISABLE_GAME_SCRIPTS - Switch scm langauge to my lua script testing.
 		TheCamera.Process();
 	}
diff --git a/src/extras/lua/lua_player.cpp b/src/extras/lua/lua_player.cpp
index 020f4cf7..c3629dd2 100644
--- a/src/extras/lua/lua_player.cpp
+++ b/src/extras/lua/lua_player.cpp
@@ -45,13 +45,17 @@ extern "C" {
 int
 createPlayerLua(lua_State *L)
 {
-	float playerIndex = (float)lua_tonumber(L, 1);
+	int playerIndex = lua_tonumber(L, 1);
 	float spawnPosX = (float)lua_tonumber(L, 2);
 	float spawnPosY = (float)lua_tonumber(L, 3);
 	float spawnPosZ = (float)lua_tonumber(L, 4);
 
-	if(!playerIndex || !spawnPosX || !spawnPosY || !spawnPosZ);
+
+	// Make sure each of the values are numbers.
+	if (!lua_isnumber(L, 1) || !lua_isnumber(L, 2) || !lua_isnumber(L, 3) || !lua_isnumber(L, 4))
+	{
 		return 0;
+	}
 
 	MiscFunctions::GetInstance().CreatePlayer(playerIndex, spawnPosX, spawnPosY, spawnPosZ);
 	
@@ -60,7 +64,7 @@ createPlayerLua(lua_State *L)
 	+ "Z: " + std::to_string(spawnPosY);
 	LogFunctions::LogInfo(Defines::logFile, logMsg);
 
-	return 4;
+	return 0;
 }
 
 /// <summary>
diff --git a/src/extras/lua/lua_test.cpp b/src/extras/lua/lua_test.cpp
index fd2c8947..5ce51ffa 100644
--- a/src/extras/lua/lua_test.cpp
+++ b/src/extras/lua/lua_test.cpp
@@ -490,7 +490,7 @@ LuaTest::LuaEvents()
 	*/
 	// luaTest.RegisterLuaFunctions(m_luaVM);
 
-	LuaTest::GetInstance().RegisterLuaFunctions(m_luaVM);	
+	// LuaTest::GetInstance().RegisterLuaFunctions(m_luaVM);	
 
 
 	// Switch to this shorter method.
@@ -663,6 +663,21 @@ extern "C" {
 	}
 }
 
+
+/**
+ * Setup the freeroam script for init, this is only called if DISABLE_GAME_SCRIPTS is toggled.
+ */
+void
+LuaTest::SetupFreeroamScript()
+{
+	// TODO Make this give a hard crash, if this isn't found there is no game scripts to load.
+	// So this should just give an error popup saying they are required.
+	if(!CheckLua(m_luaVM, luaL_dofile(m_luaVM, defines.luaFreeroamScript.c_str()))) 
+	{
+		return;
+	}
+}
+
 /// <summary>
 /// Taken idea from this in CLuaMain.cpp within MTA SA.
 /// https://github.com/multitheftauto/mtasa-blue/blob/master/Client/mods/deathmatch/logic/lua/CLuaMain.cpp
@@ -835,6 +850,16 @@ LuaTest::InitLua()
 	// First lua file
 	//-------------------
 
+	// lua_getglobal(m_luaVM, "create_player");
+	// https://stackoverflow.com/questions/76605585/calling-lua-function-from-c-minimal-example-results-in-lua-errrun#76607016
+	// TODO Test this, should get the create_player function on init.
+	// int result;
+	// result = lua_pcall(m_luaVM, 0, 0, 0);
+
+	// if (result != LUA_OK) {
+    // 	fprintf(stderr, "Error running lua: %s\n", lua_tostring(m_luaVM, -1));
+	// }
+
 	// TODO Figure out how to get this.
 	// For now I'll just use the toggles above.
 	//lua_getglobal(m_luaVM, "lose_weapons_bool");
@@ -876,6 +901,8 @@ LuaTest::InitLua()
 		// LogFunctions::LogInfo(defines.luaLogFile, toggleRespawnString);
 	}
 
+	// if(lua_isinteger(m_luaVM, -1))
+
 	// Run a function in lua
 
 	// Disabled this for now, it will be useful for later.
diff --git a/src/extras/lua/lua_test.h b/src/extras/lua/lua_test.h
index ee3b22c3..40b6ec95 100644
--- a/src/extras/lua/lua_test.h
+++ b/src/extras/lua/lua_test.h
@@ -85,6 +85,7 @@ public:
 	float vehicleTurnMass;
 
 	LuaTest();
+	void SetupFreeroamScript();
 	void InitLua();
 	void OnTick();
 	
diff --git a/src/extras/utils/defines.cpp b/src/extras/utils/defines.cpp
index aa249622..0d0941cd 100644
--- a/src/extras/utils/defines.cpp
+++ b/src/extras/utils/defines.cpp
@@ -86,7 +86,7 @@ std::string Defines::modificationAuthor = "kelson8";
 // TODO Make this grab from my latest internal gitea releases or something.
 // std::string Defines::programVersion = "1.1";
 #if defined(GTAVC)
-std::string Defines::programVersion = "1.2.5a";
+std::string Defines::programVersion = "1.2.6a";
 #elif defined(GTA3)
 std::string Defines::programVersion = "1.2a - Pre Release";
 #elif defined(GTALCS)
@@ -104,6 +104,9 @@ std::string luaScriptsFolder = Defines::gamePathFolder + "lua_scripts/";
 // Renamed from test.lua to kcnet-init.lua.
 std::string Defines::luaScript = luaScriptsFolder + "kcnet-init.lua";
 
+// New for freeroam script
+std::string Defines::luaFreeroamScript = luaScriptsFolder + "freeroam-game.lua";
+
 // New lua items
 std::string Defines::luaKeyBindScript = luaScriptsFolder + "kcnet-keybind-events.lua";
 std::string Defines::luaWastedScript = luaScriptsFolder + "kcnet-wasted.lua";
diff --git a/src/extras/utils/defines.h b/src/extras/utils/defines.h
index 4cc8c965..369d9a84 100644
--- a/src/extras/utils/defines.h
+++ b/src/extras/utils/defines.h
@@ -64,6 +64,7 @@ public:
 	static std::string fontFile;
 	// Lua
 	static std::string luaScript;
+	static std::string luaFreeroamScript;
 	static std::string luaKeyBindScript;
 	static std::string luaWastedScript;
 	static std::string luaBustedScript;
diff --git a/src/skel/glfw/glfw.cpp b/src/skel/glfw/glfw.cpp
index b7835a59..ff43d17c 100644
--- a/src/skel/glfw/glfw.cpp
+++ b/src/skel/glfw/glfw.cpp
@@ -2013,6 +2013,17 @@ GameStateStartup()
 
 #endif
 
+#if defined(LUA_TEST) && !defined(DISABLE_GAME_SCRIPTS)
+	// First, check if lua is enabled in the ini, if not run the lua init.
+	LuaTest luaTest = LuaTest();
+	if(gbLuaTest) {
+		// if(!luaInit) {
+		luaTest.InitLua();
+			// luaInit = true;
+		// }
+	}
+#endif // LUA_TEST && !DISABLE_GAME_SCRIPTS
+
 #ifdef CUSTOM_CODE
 	// Create the log folder if it doesn't exist.
 	LogFunctions::CreateLogFolder();
-- 
2.43.0

